### PR TITLE
Update Error Messaging tests and pep8 format

### PR DESCRIPTION
**Pull request by agrant23**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,3 +2,7 @@
 __pycache__
 Secure.py
 sandbox.py
+.env
```
Comment (agrant23) [inline .gitignore:5]: I did the below three lines to ignore that one file and none worked. Ugh!

File: Tools.py [hunk 2]
```diff
@@ -1,48 +1,80 @@
...
-        clean_string = re.sub("|".join(excluded_chars),"",string.printable)
-        random_string = ''.join(random.choice(clean_string) for i in range(string_len))
-        return random_string
+# String Manipulation
 
+def generate_random_string(string_len, excluded_chars=str(None)):
+    clean_string = re.sub("|".join(excluded_chars), "", string.printable)
+    random_string = ''.join(random.choice(clean_string)
+                            for i in range(string_len))
+    return random_string
```
Comment (agrant23) [inline Tools.py:21]: Below looks strange because I moved around my unused tools

File: Test_Cases.py [hunk 3]
```diff
@@ -1,169 +1,216 @@
...
+
+        acceptance criteria
+        --------------------
+        -When a password that is between 9 and 10 characters is entered,
+         the data error status and the error text exhibit's this is a
+         moderate length for the password.
+        """
+        yahoo_page = self.yahoo_page
+
+        yahoo_page.input_new_password_field(self.random_password(9))
```
Comment (agrant23) [inline Test_Cases.py:77]: The long comment below and another is my main worry with your thoughts on it's content and length.

File: Test_Cases.py [hunk 4]
```diff
@@ -1,169 +1,216 @@
...
+        acceptance criteria
+        --------------------
+        -When an 8 character password is entered, the data error status
+         and the error text exhibit's this is too short of a password.
+
+        Note
+        --------------------
+        -Yahoo security will randomly ask to "prove you are not a robot".
+         Which, at times, can be responsible for the tests in Error Message
+         Passwords class to error. There is not an easy fix for this.
```
Comment (agrant23) [inline Test_Cases.py:52]: the tests in the Error Message Passwords.

File: Test_Cases.py [hunk 5]
```diff
@@ -1,169 +1,216 @@
...
+        --------------------
+        -When a password that is between 9 and 10 characters is entered,
+         the data error status and the error text exhibit's this is a
+         moderate length for the password.
+        """
+        yahoo_page = self.yahoo_page
+
+        yahoo_page.input_new_password_field(self.random_password(9))
+        time.sleep(0.75)  # Please look at the branch Explicit_Wait_ErrorMessagePasswords if you want to see why I used this implicit wait. This keeps my code clean and readable. No E.C.s exist that can wait for this dynamic element correctly for this specific test. So I created my own E.C; which you can see being used in the given branch. This E.C. is also in Tools in this repo.
+
```
Comment (eabedrapo) [inline Test_Cases.py:79]: I'm not familiar with the part of Yahoo you're automating, but I just spent some time playing with my own account and I think I found the text you're looking to validate.

If that's the place, I noticed that when entering a more complex password, there is an element whose class changes dynamically based on the strength of the password.

So, maybe instead of using this static sleep, you should make something like:
`wait.until(EC.visibility_of_element_located((By.CSS_SELECTOR, ".error-moderate-level")))`

It may not work, and instead of using `.error-moderate-level` you may need to use a partial attribute (something like `[error-moderate-level]`, but to my point, waiting for this element to be visible using what's really changing in the code means your code will continue only when that event occurs and not when an artificial timer finishes.

Static waits such as `sleep` are a very bad practice because you artificially insert pauses that don't fix the root cause of the problem, so in one hand if the internet is slower than usual your script fails anyway, and in the other, it just keeps waiting even when the event that you're waiting to occur has already happened milliseconds ago. 

When you have multiple tests with these "sleeps" your suite becomes "flaky" and "slow", so you should always try to avoid them by any means.

File: Test_Cases.py [hunk 6]
```diff
@@ -1,169 +1,216 @@
...
+        acceptance criteria
+        --------------------
+        -When a password that is 11 characters is entered, the data
+         error and error text either does not exist or it exhibits this
+         is a password of sufficient length.
+        """
+        yahoo_page = self.yahoo_page
+
+        yahoo_page.input_new_password_field(self.random_password(11))
+        time.sleep(1)  # The Explicit_Wait_ErrorMessagePasswords branch has some really cool stuff in it. But if you look at it you will see why I decided to use this implicit wait instead
```
Comment (eabedrapo) [inline Test_Cases.py:100]: Ditto in line 78.

I just want to add that even if this line of code is meant to be a comment that you're going to delete in the future, try to keep it shorter than 80 or 100 characters (depending on the company you end working for, they might even allow lines up to 110 chars long). This is to prevent these long lines showing in multiple lines in Github, making more difficult peer reviews and horizontal scrolling when reading your code in the local IDE 😉 

You could always have a multi-line comment in the code, such as the one you did in line 110-112.

File: Test_Cases.py [hunk 7]
```diff
@@ -1,169 +1,216 @@
...
+
+        self.assertEqual(yahoo_page.long_password_error_message().
+                         get_attribute('data-error'), "STRONG_PASSWORD")
+        self.assertTrue(yahoo_page.long_password_error_message().text == "" or
```
Comment (eabedrapo) [inline Test_Cases.py:104]: You should have only one assert statement per test, regardless of the corner cases you'd like to test after the first assert. If you really need to validate all those clauses in this test, then use only `assertTrue` that validates that all your conditions have been met.

That said, a good practice for automated testing is to avoid tests that assert too many things... each test should be atomic to avoid failures to a whole test because one of the conditions in the assertion failed.

**Review by eabedrapo — COMMENTED** [review]
I tried to follow this PR, but I have to admit it's huge and I'm not familiar with the whole project that's already being worked out for a while.

I left a few comments that I hope help you to resolve the problem with waiting for the element you expect to see to validate things in your tests, and a few general recommendations about automated tests that I think would help you when start collaborating with teams professionally.

I'm not an expert, but I learned those practices from other peer reviews.

Finally, another practice when collaborating with others and requesting peer reviews:
ALWAYS try to keep Pull Requests short. As a rule of thumb, think on the goal of the PR and commit the code that's only relevant to that goal. If you've been working on too many things, separate your changes into different PRs so it will be easier to review them.

Remember that your peer reviewers will be your colleagues and in many companies are tech leads who are usually very busy, so the easier to review the code, the easier for them to give you a "Thumbs-up".

I won't approve or reject this code... as I said above, I'm not familiar of what is this doing and I'm sure you've put a lot of effort in this repo, so whether merging this PR or not, is more up to you 😉 

Enjoy the weekend!

**Comment by agrant23** [comment]
Elisa,

Thanks for the feedback and thanks for your notes on appropriate PRs; that makes sense. In case you are curious that method you mentioned to replace that time.sleep does work. Except I used Invisibility on a previous attribute value and not visibility on a current value. I wasn't expecting such a quick reply to my PR request, I should have mentioned previously I had done that, sorry. I have multiple general automation questions and questions on what should or shouldn't be in a portfolio repository. But I don't know if these specific questions may be outside of your experience. The wording of and fully explaining these questions is a little cumbersome. So, if you'd like to live chat sometime that might be best, or I can write them out; whichever you prefer.

Sorry to barrage you like this on the first day you replied to me 😳. Don't worry I won't email you this often regularly. Again thanks for your feedback.

Cordially,
Tony


________________________________

**Comment by eabedrapo** [comment]
HI @agrant23, no worries, I was in front of my computer and I wanted to help 😉 I spent a long time replacing the "sleeps" we had in the code in my previous job, so I learned to identify the changes in the page triggered by the event I want to wait to continue the execution.

Today I'm going to be busy during the day, and on weekdays I get too tired after work to join any calls... but if you want, we can organize something on Saturday.

Now, here are a few communities that I guess would help you to get connected with other people with different Selenium levels and testing communities:
* Chicago Seleniumistas discord server: https://discord.gg/C4MUDSU
* Chicago Seleniumistas meetup: https://www.meetup.com/Chicago-Seleniumistas  (I was co-organizing this group during 2019, but in Sept last year I got a new job and I couldn't help much anymore, so now Mieko and Zach are doing all the work)
* Ministry of Testing (Slack): https://www.ministryoftesting.com/slack_invite
* Techwell Communities (several communities around Agile and Testing): https://www.techwell.com/software-communities

Have a great week!
